File: java/java/src/main/java/org/signal/zkgroup/auth/AuthCredentialPresentation.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup.auth;

import java.nio.ByteBuffer;
import org.signal.zkgroup.InvalidInputException;
import org.signal.zkgroup.groups.UuidCiphertext;
import org.signal.zkgroup.internal.ByteArray;
import org.signal.client.internal.Native;

public final class AuthCredentialPresentation extends ByteArray {

  public AuthCredentialPresentation(byte[] contents) throws InvalidInputException {
    super(contents);
    try {
      Native.AuthCredentialPresentation_CheckValidContents(contents);
    } catch (IllegalArgumentException e) {
      throw new InvalidInputException(e.getMessage());
    }
  }

  public UuidCiphertext getUuidCiphertext() {
    byte[] newContents = Native.AuthCredentialPresentation_GetUuidCiphertext(contents);

    try {
      return new UuidCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public int getRedemptionTime() {
    return Native.AuthCredentialPresentation_GetRedemptionTime(contents);
  }

}

File: swift/Sources/SignalClient/zkgroup/AuthCredentialPresentation.swift
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class AuthCredentialPresentation: ByteArray {

  public required init(contents: [UInt8]) throws {
    try super.init(contents, checkValid: signal_auth_credential_presentation_check_valid_contents)
  }

  public func getUuidCiphertext() throws -> UuidCiphertext {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningSerialized {
        signal_auth_credential_presentation_get_uuid_ciphertext($0, contents)
      }
    }
  }

  public func getRedemptionTime() throws -> UInt32 {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningInteger {
        signal_auth_credential_presentation_get_redemption_time($0, contents)
      }
    }
  }

}
